### PR TITLE
Docker + Upgrade documentation

### PR DESCRIPTION
**Pull request by cam-schultz**
## Why this should be merged
Adds documentation for:
~~1. Explaining the steps to deploy Teleporter in the Docker setup, mapping to `run_setup.sh`~~
2. Performing a Teleporter version upgrade with a node and a relayer

The upgrade documentation outlines the necessary steps at a high level; it does not provide detailed instructions on how to accomplish those steps. For example, managing a node's chain config is omitted here, as it is documented elsewhere.

## How this works
N/A

## How this was tested
N/A

## How is this documented
Self explanatory.

### CODE REVIEW

File: contracts/src/Teleporter/upgrades/UPGRADING.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+## Register a new Teleporter version
+
+Once a new Teleporter contract instance is [deployed](../../../../utils/contract-deployment/README.md), the `addProtocolVersion` method of [TeleporterRegistry.sol](./TeleporterRegistry.sol) must be called. This method is only callable if the associated Warp message was sent via an off-chain Warp message, which is provided by a validator's chain config.
+
+The steps to do so are as follows:
+
+1. Construct the unsigned Warp message bytes:
+
+    a. Pack the Warp payload to be parsed by `addProtocolVersion`. This is a tuple of `(ProtocolRegistryEntry, address)`, where the `ProtocolRegistryEntry` specifies the new `TeleporterMessenger` contract address and the version, and the `address` specifies the `TeleporterRegistry` contract address that the new Teleporter version will be registered with
```
Comment (geoff-vball) [inline contracts/src/Teleporter/upgrades/UPGRADING.md:13]:
```suggestion
    a. Pack the Warp payload to be parsed by `addProtocolVersion`. This is a tuple of `(ProtocolRegistryEntry, address)`, where the `ProtocolRegistryEntry` specifies the new `TeleporterMessenger` contract address and the version, and the `address` specifies the `TeleporterRegistry` contract address that the new Teleporter version will be registered with.
```

File: contracts/src/Teleporter/upgrades/UPGRADING.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    b. Pack the Warp payload as an [AddressedCall](https://github.com/ava-labs/avalanchego/blob/v1.11.0-fuji/vms/platformvm/warp/payload/addressed_call.go#L15), with the source address set to the zero address. This is how `addProtocolVersion` identifies this message as an off-chain Warp message.
+    
+    c. Pack the `AddressedCall` message into an [unsigned Warp message](https://github.com/ava-labs/avalanchego/blob/v1.11.0-fuji/vms/platformvm/warp/unsigned_message.go#L14), specifying the blockchain ID that `TeleporterRegistry` and the new `TeleporterMessenger` are deployed to.
+
+2. Populate the "warp-off-chain-messages" field of each validator node's chain config with the hex-encoded unsigned Warp message bytes
+
+3. Restart the node to mark the off-chain Warp message as eligible for signing by the validator
```
Comment (geoff-vball) [inline contracts/src/Teleporter/upgrades/UPGRADING.md:21]:
```suggestion
2. Populate the "warp-off-chain-messages" field of each validator node's chain config with the hex-encoded unsigned Warp message bytes.

3. Restart the node to mark the off-chain Warp message as eligible for signing by the validator.
```

File: docker/README.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+```bash
+avalanche subnet create subneta --force --genesis ./subnetGenesis_A.json --config ./docker/defaultNodeConfig.json --evm --vm-version $SUBNET_EVM_VERSION --log-level info --skip-update-check
+avalanche subnet configure subneta --node-config ./docker/defaultNodeConfig.json --chain-config ./docker/defaultChainConfig.json --skip-update-check
+avalanche subnet deploy subneta --local --avalanchego-version $AVALANCHEGO_VERSION --config ./docker/defaultNodeConfig.json --log-level info --skip-update-check
+```
+
+The script then derives the Teleporter deployment information using the [contract utility](../utils/contract-deployment/contractDeploymentTools.go) included in this repository:
+```bash
+go run utils/contract-deployment/contractDeploymentTools.go constructKeylessTx contracts/out/TeleporterMessenger.sol/TeleporterMessenger.json
+teleporter_deploy_address=$(cat UniversalTeleporterDeployerAddress.txt)
```
Comment (geoff-vball) [inline docker/README.md:24]:
```suggestion
teleporter_deployer_address=$(cat UniversalTeleporterDeployerAddress.txt)
```

File: contracts/src/Teleporter/upgrades/UPGRADING.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
+# Upgrading Teleporter 
+
+This document outlines the steps necessary to upgrade Teleporter as a subnet operator, a relayer operator, and a dApp admin. These steps are intentionally high-level, and leave the implementation details unspecified. As a reference, the [Teleporter Registry test suite](../../../../tests/flows/teleporter_registry.go) implements the steps described in a test environment.
+
+## Register a new Teleporter version
```
Comment (meaghanfitzgerald) [inline contracts/src/Teleporter/upgrades/UPGRADING.md:5]:
```suggestion
## Register a New Teleporter Version
```

File: contracts/src/Teleporter/upgrades/UPGRADING.md [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    
+    b. Set the value as the following, populating the <TELEPORTER_REGISTRY_ADDRESS>:
+    ```json
+    {
+        "message-format": "off-chain-registry",
+        "settings": {
+            "teleporter-registry-address": "<TELEPORTER_REGISTRY_ADDRESS>"
+        }
+    }
+    ```
```
Comment (meaghanfitzgerald) [inline contracts/src/Teleporter/upgrades/UPGRADING.md:47]:
```suggestion
    b. Set the value as the following, populating the TELEPORTER_REGISTRY_ADDRESS:
    ```json
    {
        "message-format": "off-chain-registry",
        "settings": {
            "teleporter-registry-address": "TELEPORTER_REGISTRY_ADDRESS"
        }
    }
    ```
```

File: contracts/src/Teleporter/upgrades/UPGRADING.md [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+1. Construct a "manual-warp-messages" entry in the AWM Relayer configuration file, using the following values:
+
+    a. "unsigned-message-bytes": the hex-encoded unsigned Warp message bytes derived above.
+
+    b. "source-blockchain-id": the blockchain ID that that `TeleporterRegistry` and the new `TeleporterMessenger` are deployed to.
+
+    c. "destination-blockchain-id": the blockchain ID that that `TeleporterRegistry` and the new `TeleporterMessenger` are deployed to.
+
+    d. "source-address": the zero address, ""0x0000000000000000000000000000000000000000". This is the "source" address for off-chain Warp messages.
```
Comment (meaghanfitzgerald) [inline contracts/src/Teleporter/upgrades/UPGRADING.md:33]:
```suggestion
    d. "source-address": the zero address, "0x0000000000000000000000000000000000000000". This is the "source" address for off-chain Warp messages.
```

File: contracts/src/Teleporter/upgrades/UPGRADING.md (hunk 5 above)
Comment (geoff-vball) [reply contracts/src/Teleporter/upgrades/UPGRADING.md:47]: @meaghanfitzgerald Do these brackets cause some sort of formatting issue? We usually use angled brackets to denote some value that needs to be filled in by the user.

File: docker/README.md (hunk 3 above)
Comment (cam-schultz) [reply docker/README.md]: Renamed this variable in all instances

File: contracts/src/Teleporter/upgrades/UPGRADING.md [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+1. Construct the unsigned Warp message bytes:
+
+    a. Pack the Warp payload to be parsed by `addProtocolVersion`. This is a tuple of `(ProtocolRegistryEntry, address)`, where the `ProtocolRegistryEntry` specifies the new `TeleporterMessenger` contract address and the version, and the `address` specifies the `TeleporterRegistry` contract address that the new Teleporter version will be registered with.
+
+    b. Pack the Warp payload as an [AddressedCall](https://github.com/ava-labs/avalanchego/blob/v1.11.0-fuji/vms/platformvm/warp/payload/addressed_call.go#L15), with the source address set to the zero address. This is how `addProtocolVersion` identifies this message as an off-chain Warp message.
+    
+    c. Pack the `AddressedCall` message into an [unsigned Warp message](https://github.com/ava-labs/avalanchego/blob/v1.11.0-fuji/vms/platformvm/warp/unsigned_message.go#L14), specifying the blockchain ID that `TeleporterRegistry` and the new `TeleporterMessenger` are deployed to.
+
+2. Populate the "warp-off-chain-messages" field of each validator node's chain config with the hex-encoded unsigned Warp message bytes.
```
Comment (minghinmatthewlam) [inline contracts/src/Teleporter/upgrades/UPGRADING.md:19]: do we need to describe steps for getting the hex encoded bytes?

File: docker/run_relayer.sh [hunk 8]
```diff
@@ -34,6 +34,13 @@ set +a
 reward_address=0xA100fF48a37cab9f87c8b5Da933DA46ea1a5fb80
 account_private_key=C2CE4E001B7585F543982A01FBC537CFF261A672FA8BD1FAFC08A207098FE2DE
 
+# Send tokens to cover gas costs for the relayers.
+cast send --private-key $user_private_key --value 500ether $relayer_address --rpc-url $subnet_a_rpc_url
```
Comment (minghinmatthewlam) [inline docker/run_relayer.sh:38]: do we need to clear up we're using avax for gas though the command specifies ether?

**Review by geoff-vball — DISMISSED** [review]

**Review by geoff-vball — APPROVED** [review]

**Review by michaelkaplan13 — APPROVED** [review]

**Pull request merged**